services/motion/builtin/state/state.go

```diff
@@ -52,16 +67,529 @@ type PlannerExecutor interface {
+	e.state.waitGroup.Add(1)
+	e.waitGroup.Add(1)
+
+	utils.PanicCapturingGo(func() {
+		defer e.state.waitGroup.Done()
+		defer e.waitGroup.Done()
+
+		lastPWE := originalPlanWithExecutor
+		// Exit conditions of this loop:
+		// 1. The execution's context was cancelled, which happens if the state's Stop() was called or
+		// StopExecutionByResource was called for this resource
+		// 2. the execution succeeded
+		// 3. the execution failed
+		// 4. replanning failed
+		for {
+			resChan := make(chan struct {
+				resp ExecuteResponse
+				err  error
+			}, 1)
+			utils.PanicCapturingGo(func() {
+				replan, err := lastPWE.planExecutor.Execute(lastPWE.waypoints)
+				resChan <- struct {
+					resp ExecuteResponse
+					err  error
+				}{replan, err}
+			})
+			select {
+			case <-e.cancelCtx.Done():
+				lastPWE.planExecutor.Cancel()
+				e.notifyStatePlanStopped(lastPWE.plan, time.Now())
+				return
+			case res := <-resChan:
+				// failure
+				if res.err != nil {
+					// cancel the context now that the context has termintaed to not leak context
+					e.cancelFunc()
+					e.notifyStatePlanFailed(lastPWE.plan, res.err.Error(), time.Now())
+					return
+				}
+
+				// success
+				if !res.resp.Replan {
+					// cancel the context now that the context has termintaed to not leak context
+					e.cancelFunc()
+					e.notifyStatePlanSucceeded(lastPWE.plan, time.Now())
+					return
+				}
+
+				replanCount++
+				newPWE, err := e.newPlanWithExecutor(lastPWE.motionplan, replanCount)
+				// replan failed
+				if err != nil {
+					msg := "failed to replan for execution %s and component: %s, " +
+						"due to replan reason: %s, tried setting previous plan %s " +
+						"to failed due to error: %s\n"
+					e.logger.Warnf(msg, e.id, e.componentName, res.resp.ReplanReason, lastPWE.plan.ID, err.Error())
+
+					e.notifyStatePlanFailed(lastPWE.plan, err.Error(), time.Now())
+					return
+				}
+
+				e.notifyStateRePlan(lastPWE.plan, res.resp.ReplanReason, newPWE.plan, time.Now())
+				lastPWE = newPWE
+			}
+		}
+	})
+
+	return nil
+}
+
+func (e *execution[R]) toStateExecution() stateExecution {
+	return stateExecution{
+		id:            e.id,
+		componentName: e.componentName,
+		waitGroup:     e.waitGroup,
+		cancelFunc:    e.cancelFunc,
+	}
+}
+
+func (e *execution[R]) notifyStateNewExecution(execution stateExecution, plan motion.Plan, time time.Time) {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	// NOTE: We hold the lock for both updateStateNewExecution & updateStateNewPlan to ensure no readers
+	// are able to see a state where the execution exists but does not have a plan with a status.
+	e.state.updateStateNewExecution(execution)
+	e.state.updateStateNewPlan(newPlanMsg{
+		plan:       plan,
+		planStatus: motion.PlanStatus{State: motion.PlanStateInProgress, Timestamp: time},
+	})
+}
+
+func (e *execution[R]) notifyStateRePlan(lastPlan motion.Plan, reason string, newPlan motion.Plan, time time.Time) {
```

Comment:
I think the p in replan should be lowercase

Reply:
done